source/fe/mapping_q.cc

```diff
@@ -1411,85 +1389,81 @@ MappingQ<dim, spacedim>::fill_mapping_data_for_generic_points(
-          for (unsigned int j = 0; j < n_lanes && i + j < n_points; ++j)
-            for (unsigned int d = 0; d < spacedim; ++d)
-              output_data.quadrature_points[i + j][d] = result.first[d][j];
+  internal::MappingQImplementation::maybe_update_q_points_Jacobians_generic(
+    CellSimilarity::none,
+    data,
+    unit_points,
+    polynomials_1d,
+    polynomial_degree,
+    renumber_lexicographic_to_hierarchic,
+    output_data.quadrature_points,
+    output_data.jacobians,
+    output_data.inverse_jacobians);
+}
 
-        if (update_flags & update_jacobians)
-          for (unsigned int j = 0; j < n_lanes && i + j < n_points; ++j)
-            for (unsigned int d = 0; d < spacedim; ++d)
-              for (unsigned int e = 0; e < dim; ++e)
-                output_data.jacobians[i + j][d][e] = result.second[e][d][j];
 
-        if (update_flags & update_inverse_jacobians)
-          {
-            DerivativeForm<1, spacedim, dim, VectorizedArray<double>> jac(
-              result.second);
-            const DerivativeForm<1, spacedim, dim, VectorizedArray<double>>
-              inv_jac = jac.covariant_form();
-            for (unsigned int j = 0; j < n_lanes && i + j < n_points; ++j)
-              for (unsigned int d = 0; d < dim; ++d)
-                for (unsigned int e = 0; e < spacedim; ++e)
-                  output_data.inverse_jacobians[i + j][d][e] = inv_jac[d][e][j];
-          }
-      }
-    else
-      {
-        const auto result =
-          internal::evaluate_tensor_product_value_and_gradient(
-            polynomials_1d,
-            support_points,
-            unit_points[i],
-            polynomial_degree == 1,
-            renumber_lexicographic_to_hierarchic);
-
-        if (update_flags & update_quadrature_points)
-          output_data.quadrature_points[i] = result.first;
-
-        if (update_flags & update_jacobians)
-          {
-            DerivativeForm<1, spacedim, dim> jac = result.second;
-            output_data.jacobians[i]             = jac.transpose();
-          }
 
-        if (update_flags & update_inverse_jacobians)
-          {
-            DerivativeForm<1, spacedim, dim> jac(result.second);
-            DerivativeForm<1, spacedim, dim> inv_jac = jac.covariant_form();
-            for (unsigned int d = 0; d < dim; ++d)
-              for (unsigned int e = 0; e < spacedim; ++e)
-                output_data.inverse_jacobians[i][d][e] = inv_jac[d][e];
-          }
-      }
+template <int dim, int spacedim>
+void
+MappingQ<dim, spacedim>::fill_mapping_data_for_face_quadrature(
+  const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+  const unsigned int                                          face_no,
+  const Quadrature<dim - 1> &                                 face_quadrature,
+  const UpdateFlags                                           update_flags,
+  internal::FEValuesImplementation::MappingRelatedData<dim, spacedim>
+    &output_data) const
```

Comment:
I don't understand why you don't create a `InternalData` within MappingInfo and fill this with the points? Are there particular reasons against this?

Reply:
I can do that but the call to `compute_mapping_support_points()` and `internal::MappingQImplementation::do_fill_fe_face_values()` (currently) need to stay in `MappingQ` because they are protected or need protected arguments. (I can make things public if we choose to do so)

Reply:
> I can do that but the call to compute_mapping_support_points() and internal::MappingQImplementation::do_fill_fe_face_values() (currently) need to stay in MappingQ because they are protected or need protected arguments. (I can make things public if we choose to do so)

I don't understand what you mean...

Reply:
`InternalData` is now filled inside `MappingInfo` and only the function call to the internal MappingQ function is realized in `fill_mapping_data_for_face_quadrature()`. This function is protected similarly to the `fill_fe_face_values()` function. `fill_mapping_data_for_face_quadrature()` does the function call with the protected arguments from MappingQ and calls the internal function `do_fill_fe_face_values()`.